src/main/java/com/openlattice/chronicle/services/ChronicleServiceImpl.java

```diff
@@ -558,91 +487,205 @@ private void updateParticipantMetadata(
+    ) {
+        Map<UUID, Map<UUID, Map<UUID, Set<Object>>>> entityKeysByEntitySet = Maps.newHashMap();
+
+        entitiesByEntityKey.forEach( ( entityKey, entity ) -> {
+            UUID entitySetId = entityKey.getEntitySetId();
+            UUID entityKeyId = entityKeyIdMap.get( entityKey );
+
+            Map<UUID, Map<UUID, Set<Object>>> mappedEntity = entityKeysByEntitySet
+                    .getOrDefault( entitySetId, Maps.newHashMap() );
+            mappedEntity.put( entityKeyId, entity );
+
+            entityKeysByEntitySet.put( entitySetId, mappedEntity );
+        } );
+
+        return entityKeysByEntitySet;
+    }
+
+    private Set<DataEdgeKey> getDataEdgeKeysFromEntityKeys(
+            Set<Triple<EntityKey, EntityKey, EntityKey>> edgesByEntityKey,
+            Map<EntityKey, UUID> entityKeyIdMap
+    ) {
+        return StreamUtil.stream( edgesByEntityKey )
+                .map( triple -> {
+                    UUID srcEKID = entityKeyIdMap.get( triple.getLeft() );
+                    UUID edgeEKID = entityKeyIdMap.get( triple.getMiddle() );
+                    UUID dstEKID = entityKeyIdMap.get( triple.getRight() );
+
+                    UUID srcESID = triple.getLeft().getEntitySetId();
+                    UUID edgeESID = triple.getMiddle().getEntitySetId();
+                    UUID dstESID = triple.getRight().getEntitySetId();
+
+                    EntityDataKey src = new EntityDataKey( srcESID, srcEKID );
+                    EntityDataKey edge = new EntityDataKey( edgeESID, edgeEKID );
+                    EntityDataKey dst = new EntityDataKey( dstESID, dstEKID );
+
+                    return new DataEdgeKey( src, dst, edge );
+                } )
+                .collect( Collectors.toSet() );
     }
 
-    private void createAppDataEntitiesAndAssociations(
+    // create entities and edges
+    private void createEntitiesAndAssociations(
             DataApi dataApi,
+            DataIntegrationApi dataIntegrationApi,
             List<SetMultimap<UUID, Object>> data,
-            UUID deviceEntityKeyId,
+            UUID studyId,
+            UUID deviceEKID,
             String participantId,
-            UUID participantEntityKeyId,
-            UUID participantEntitySetId,
-            UUID studyId
+            String deviceId,
+            UUID participantEKID,
+            UUID participantESID
     ) {
+        ListMultimap<UUID, Map<UUID, Set<Object>>> appDataEntities = ArrayListMultimap.create();
+        ListMultimap<UUID, DataAssociation> appDataAssociations = ArrayListMultimap.create();
 
-        Stopwatch stopwatch = Stopwatch.createStarted();
-        ListMultimap<UUID, Map<UUID, Set<Object>>> entities = ArrayListMultimap.create();
-        ListMultimap<UUID, DataAssociation> associations = ArrayListMultimap.create();
+        Map<EntityKey, Map<UUID, Set<Object>>> entitiesByEntityKey = Maps.newHashMap();
+        Set<Triple<EntityKey, EntityKey, EntityKey>> edgesByEntityKey = Sets.newHashSet();
```

Comment:
It might be helpful for future development to include a comment indicating the order of the triple here (i.e. <src, association, dst>, <src, dst, association>, etc)